### PR TITLE
Feature parity of 3.5

### PR DESCRIPTION
**Pull request by joelgomes1994**
## Overview
- Implement missing features to Android GPGS native plugin.
- Changes based on the [4.x](https://github.com/Iakobs/godot-play-game-services) plugin repository.
- Related to #3 and [this PR](https://github.com/Iakobs/godot-google-play-game-services-plugin/pull/5).

## Implementation
- [x] [Snapshots](https://developers.google.com/android/reference/com/google/android/gms/games/SnapshotsClient) (for saving and loading games)
- [x] [Events](https://developers.google.com/android/reference/com/google/android/gms/games/EventsClient) (for tracking and statistics)
- [x] Implement missing methods for existing features
- [x] Rewrite README

## Fixes
- [x] Add [Gson](https://mvnrepository.com/artifact/com.google.code.gson/gson) dependency to `GodotGooglePlayGameServices.gdap`
    - Fixes runtime exceptions/crashes when serializing return values
- [x] `submitScore's` `score` is a `Float` argument, not `Int`

## Project
- Remove `godot-lib.3.5.2.stable.release.aar` from versioning, as it's a bad practice to add lib binaries to the versioning. Add manually the desired lib version to the folder `godot-lib.release`
- Reorganize folders structure of source code

**Binaries:** [GodotGooglePlayGameServices.zip](https://github.com/Iakobs/godot-google-play-game-services-android-plugin/files/13880728/GodotGooglePlayGameServices.zip)

## Example video
https://www.youtube.com/watch?v=RbEh7WLQ_d8

### CODE REVIEW

**Comment by Iakobs** [comment]
Hey @joelgomes1994 I'm commenting here, but I've checked both PRs.

First of all, thanks a lot!!! This is amazing work and SUPER helpful!

For me everything looks good. I love how you reorganised the godot plugin in the other PR. I was also thinking about putting everything in a single autload for the Godot 4.2 version, to not clutter the namespace, but I also hate breaking the single responsability principle so I had an internal battle and the second option won 😝 

The only thing preventing me from approving the PR right now is the Events feature from google play. I know that's not telemetry _per se_, but I wasn't planning on adding that feature to the plugin because I, personally, don't like the idea behind it.

Let me think about that, because I'm also aware that tracking the player's interactions with the game is also a very important feature for games, so I don't want to decide lightly on removing such a feature from it. I will get back to you on both PRs ASAP.

Again, thanks a lot for your work! Have a very happy new year!!

**Comment by joelgomes1994** [comment]
I'm glad you found it helpful, I was wondering if this many changes would be too much for your liking :laughing:.

I'm just updating the README instructions to be more clear, after that I'll change the PR status from draft to ready to ready for review.

About the Events it's ok, although I think that it's in the hands of the end-user to use it or not, if you judge to not include into the plugin, I'll remove it in no time.

File: app/src/main/java/com/jacobibanez/godot/gpgs/signals/Signals.kt [hunk 1]
```diff
@@ -179,16 +178,16 @@ object SnapshotSignals {
      * This signal is emitted when calling the [com.jacobibanez.godot.gpgs.GodotGooglePlayGameServices.loadGame] method
      * or after selecting a saved game in the window opened by the [com.jacobibanez.godot.gpgs.GodotGooglePlayGameServices.showSavedGames] method.
      *
-     * @return A [Dictionary] representing a [com.google.android.gms.games.snapshot.Snapshot](https://developers.google.com/android/reference/com/google/android/gms/games/snapshot/Snapshot).
+     * @return A JSON representing a [com.google.android.gms.games.snapshot.Snapshot](https://developers.google.com/android/reference/com/google/android/gms/games/snapshot/Snapshot).
      */
-    val gameLoaded = SignalInfo("gameLoaded", Dictionary::class.java)
+    val gameLoaded = SignalInfo("gameLoaded", String::class.java)
```
Comment (Iakobs) [inline app/src/main/java/com/jacobibanez/godot/gpgs/signals/Signals.kt]: I understand you did this for consistency with the other signals, but I was thinking on removing the GSON dependency and using dictionaries and arrays instead, what do you think? this way we can remove a 3rd party library.

File: README.md [hunk 2]
```diff
@@ -1,117 +1,138 @@
 ![Plugin Header](github-social-preview.png)
 
 [![Android version and API level](https://img.shields.io/badge/Android-API%20Level%2033-darkgreen.svg)](https://developer.android.com)
-[![Godot version](https://img.shields.io/badge/Godot%20Engine-3.5.2-blue.svg)](https://github.com/godotengine/godot/)
+[![Godot version](https://img.shields.io/badge/Godot%20Engine-3.5.3-blue.svg)](https://github.com/godotengine/godot/)
 [![Google Play Game Services version](https://img.shields.io/badge/Play%20Games%20Services%20v2-17.0.0-green.svg)](https://developers.google.com/games/services/android/quickstart)
 
-> ⚠️ A new version of this plugin for Godot 4.2 can be found [here](https://github.com/Iakobs/godot-play-game-services)!
+> ⚠️ A **Godot 4.x** version of this plugin can be found [here](https://github.com/Iakobs/godot-play-game-services)!
```
Comment (Iakobs) [inline README.md]: Same comment as in the other PR :)

File: app/src/main/java/com/jacobibanez/godot/gpgs/signals/Signals.kt (hunk 1 above)
Comment (joelgomes1994) [reply app/src/main/java/com/jacobibanez/godot/gpgs/signals/Signals.kt]: It makes sense! However, the [Godot docs for 3.5 Android plugins](https://docs.godotengine.org/en/3.5/tutorials/platform/android/android_plugin.html#troubleshooting) states that only specific array types can be used, so I'm unsure if arrays of dictionaries will work, but I'll try.

image

File: README.md (hunk 2 above)
Comment (joelgomes1994) [reply README.md]: Ok!

File: app/src/main/java/com/jacobibanez/godot/gpgs/signals/Signals.kt (hunk 1 above)
Comment (joelgomes1994) [reply app/src/main/java/com/jacobibanez/godot/gpgs/signals/Signals.kt]: Yeah, it seems that Godot 3.x does not support emitting other than basic types arrays, as the documentation stated. The value becomes `null`.

```
E 0:04:24.886   emit_signal: Error calling method from signal 'eventsLoaded': 'Node(singleton.gd)::_on_events_loaded': Cannot convert argument 1 from Nil to Array..
  <C++ Source>  core/object.cpp:1242 @ emit_signal()
```

**Edit:** to clarify, for dictionaries it works, but for arrays of dictionaries it doesn't. I think that, to make the return values more uniform, dictionaries and arrays should be serialized to JSON strings, but if you prefer the arrays to be serialized to strings and dictionaries be kept as dictionaries, I can change them back.

Comment (Iakobs) [reply app/src/main/java/com/jacobibanez/godot/gpgs/signals/Signals.kt]: oooooh! I see!! Let's keep it all JSON strings then, I agree this way is more uniform!

**Review by Iakobs — DISMISSED** [review]

**Review by Iakobs — DISMISSED** [review]

**Comment by joelgomes1994** [comment]
Updated the reference docs, sorry. One last review and I can merge it. I commented [this](https://github.com/Iakobs/godot-google-play-game-services-plugin/pull/5#issuecomment-1883892188) in the other repository.

**Review by Iakobs — APPROVED** [review]

**Comment by joelgomes1994** [comment]
@Iakobs As you have seen, I already built and published the [latest release](https://github.com/Iakobs/godot-google-play-game-services-android-plugin/releases) binary, so now it's up to you to publish the native plugin and [GDScript wrapper](https://github.com/Iakobs/godot-google-play-game-services-plugin) to the Asset Library. Thanks for the trust to make these changes!

**Comment by Iakobs** [comment]
yes! and thanks to you for all the hard work!!!! I'm also publishing a new release of the gdscript wrapper!!

**Pull request merged**